### ISSUE TITLE
Elastic distributed training with Lightning

### ISSUE THREAD

**Issue by thecooltechguy** [PyTorchLightning/pytorch-lightning#9063]
## 🚀 Feature

This is a feature request to be able to run distributed training jobs with Lightning, where the number of nodes may increase/decrease over time. For example, a distributed training job could start off with 1 node, and then more nodes could be added to the cluster later, at which point the job would include those nodes in the training process.

### Motivation

We do a lot of distributed training (involving 4-8 p4d.24xlarge/p3dn.24xlarge EC2 instances on AWS), and due to these gpu servers being in-demand for cloud distributed training in general, we very frequently face EC2 insufficient capacity issues when we try to launch an node group of 4-8 such servers within the same AZ. This applies for on-demand instances, and is even more true for spot instances.

So, typically, we'll only be able to immediately launch 1-2 GPU nodes, with the rest becoming available hours (or even 1+ days) later. If Lightning can do elastic distributed training, however, we can immediately start the job with the 1-2 servers that are immediately available, and then _**resume**_ the training job as more nodes become available and join the cluster.

When using spot instances, another motivation point would be to add fault-tolerance to the training job, so that the job can continue as spot instances come/go throughout the training job, greatly reducing the cost of training (~50%+ savings, depending on the AWS region). For example, we could have 1 on-demand instance and multiple spot instances per job, to be much more cost efficient due to fault-tolerance from elastic training. Less training iteration cost --> more iterations --> faster results!

### Pitch

Ideally, it would be cool if we can specify `min_nodes` and `max_nodes` parameters in the Lightning Trainer (which are only valid when the `ClusterEnvironment` is based on multiple nodes, for example), and then launch training in Kubernetes/SLURM/etc. 

Then, when the node count changes, I think this can be handled a few different ways. I'm not an expert here since I don't have any experience with elastic, distributed training, but some example projects that handle this are Horovod Elastic and torchelastic. I think torchelastic will simply stop the job and restart the training pods on k8s once there's a node count change, and then our training script would have to do any necessary check-pointing/hyper-parameter/configuration changes (such as learning rate tuning, dataloader distributed sampling with the new set of nodes, etc.).

### Alternatives

I'm not sure if there are any real, viable alternatives to elastic training, apart from just waiting for all node resources to become available before training begins (which can also be expensive).

### Additional context

In case this info helps, currently we're extensively using Lightning to run distributed training jobs with DDP and torchelastic on k8s, on top of AWS EKS, using p4d.24xlarge on-demand and (occasionally) spot instances.

Also, I'm not sure if/how elastic training will affect model convergence. I would love to hear any thoughts the Lightning team may have about this w.r.t how elastic training affects model convergence, based on your experience!

And finally, Lightning is AWESOME!!
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by yifuwang** [comment]
@kiukchung - any recommendations?

**Comment by kiukchung** [comment]
torchelastic indeed enables what is described above. TL;DR torchelastic makes the training job "as elastic as possible" without requiring a ton of changes to the training script / pytorch. Just make sure you checkpoint as often as you can tolerate scaling events and reload from the most-recent checkpoint. We've tried the finer grained elastic trainers and in practice they not only do not work but they are hard to maintain and improve over time due to the user having to be uber conscious about state management.

To actually understand what is needed to support "elasticity" we have to dissect it into a few parts:

1. Node (or worker) membership changes (departures and admission)
2. DDP being able to update ranks on such changes
3. Your trainer (or model) being able to correct adapt to changes in world_size

**1. Node (or worker) membership changes (departures and admission)**
torchelastic deals with this. As @thecooltechguy correctly pointed out, when a node/worker joins/leaves torchelastic will first check whether the job is still `(min - max)` compliant and will restart ALL the worker procs.

**2. DDP being able to update ranks on such changes**
the reason why torchelastic restarts all the worker procs on scaling events is because DDP's world_size is static. We can try fixing this in two ways:

1. re-initialize process groups inplace on surviving procs (and this is precisely what torchelastic v0 did) but that comes with a TON of other baggage (e.g. there are other states outside of process groups that also need to be restored or reinitialized and it is hard to keep track of all of them, especially so if they are user-defined).
2. make DDP (process groups) elastic - this is non-trivial since it also implies you have to make the backends (nccl, gloo) elastic

**3. Your trainer (or model) being able to correct adapt to changes in world_size**
In the simplest case your global batch size changes when the world_size changes, and you could deal with this by adjusting the learning rate and other hyperparameters that are a function of global batch size accordingly. But even if the results are good (unless we keep track of scaling events) it will be impossible to recreate the experiments unless we reproduce the scaling events w.r.t the training progress)

**Comment by thecooltechguy** [comment]
Hi @kiukchung thanks for sharing this! It helped me better understand how torchelastic does elastic training.

So, is there a way to use Lightning while also doing elastic training with torchelastic? 

Resuming from the latest checkpoint upon a restart won't be an issue I think, but I'm wondering how we could change things like the `num_nodes` parameter of the Trainer upon a torchelastic restart, update learning rate, etc.

Would love any pointers for this, thanks! :)

**Comment by kiukchung** [comment]
@thecooltechguy happy to help/provide context! Assuming you have the following:

1. some code in your training script that loads from "the-most-recent-checkpoint"
2. no special ddp (process_group) initialization (e.g. that you are just going through lightning to get this done).

Things should work for you almost out-of-the-box. Check out this docs page: https://pytorch-lightning.readthedocs.io/en/stable/advanced/multi_gpu.html#torch-distributed-elastic

torchelastic is exposed to the end-user as a CLI (if you've used MPI think of it as `mpirun`) so its completely decoupled from your training script. More info on how torchelastic works can be found here: https://pytorch.org/docs/stable/elastic/agent.html

Also, if you could share what your scheduler/infra setup is, I can point you to relevant docs/best-practices.
